components/sensor/bme280/bme280.go

```diff
@@ -0,0 +1,570 @@
+		func(attributes config.AttributeMap) (interface{}, error) {
+			var conf AttrConfig
+			return config.TransformAttributeMapToStruct(&conf, attributes)
+		}, &AttrConfig{})
+}
+
+func newSensor(
+	ctx context.Context,
+	deps registry.Dependencies,
+	name string,
+	attr *AttrConfig,
+	logger golog.Logger,
+) (sensor.Sensor, error) {
+	b, err := board.FromDependencies(deps, attr.Board)
+	if err != nil {
+		return nil, fmt.Errorf("bme280 init: failed to find board: %w", err)
+	}
+	localB, ok := b.(board.LocalBoard)
+	if !ok {
+		return nil, fmt.Errorf("board %s is not local", attr.Board)
+	}
+	i2cbus, ok := localB.I2CByName(attr.I2CAttrConfig.I2CBus)
+	if !ok {
+		return nil, fmt.Errorf("bme280 init: failed to find i2c bus %s", attr.I2CAttrConfig.I2CBus)
+	}
+	addr := attr.I2CAttrConfig.I2cAddr
+	if addr == 0 {
+		addr = defaultI2Caddr
+		logger.Warn("using i2c address : 0x77")
+	}
+	baudrate := attr.I2CAttrConfig.I2CBaudRate
+	if baudrate == 0 {
+		baudrate = defaultBaud
+		logger.Warn("using default baudrate : 100000")
+	}
+
+	s := &bme280{
+		name:     name,
+		logger:   logger,
+		bus:      i2cbus,
+		addr:     byte(addr),
+		wbaud:    baudrate,
+		lastTemp: -999, // initialize to impossible temp
+	}
+
+	err = s.reset(ctx)
+	if err != nil {
+		return nil, err
+	}
+	time.Sleep(50 * time.Millisecond)
+	s.calibration = map[string]int{}
+	err = s.setupCalibration(ctx)
+	if err != nil {
+		return nil, err
+	}
+	err = s.setMode(ctx, 0b11)
+	if err != nil {
+		return nil, err
+	}
+
+	err = s.setStandbyTime(ctx, 0)
+	if err != nil {
+		return nil, err
+	}
+	err = s.setFilter(ctx, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	// humidity
+	err = s.setOverSample(ctx, bme280CTRLHumidityReg, 0, 1) // Default of 1x oversample
+	if err != nil {
+		return nil, err
+	}
+	// temperature
+	err = s.setOverSample(ctx, bme280CTRLMEASReg, 5, 1) // Default of 1x oversample
+	if err != nil {
+		return nil, err
+	}
+	// pressure
+	err = s.setOverSample(ctx, bme280CTRLMEASReg, 2, 1) // Default of 1x oversample
+	if err != nil {
+		return nil, err
+	}
+
+	return s, nil
+}
+
+// bme280 is a i2c sensor device.
+type bme280 struct {
+	generic.Unimplemented
+	logger golog.Logger
+
+	bus         board.I2C
+	addr        byte
+	wbaud       int
+	name        string
+	calibration map[string]int
+	lastTemp    float64 // Store raw data from temp for humidity calculations
+}
+
+// Readings returns a list containing single item (current temperature).
+func (s *bme280) Readings(ctx context.Context) (map[string]interface{}, error) {
+	handle, err := s.bus.OpenHandle(s.addr)
+	if err != nil {
+		s.logger.Errorf("can't open bme280 i2c %s", err)
+		return nil, err
+	}
+	err = handle.Write(ctx, []byte{byte(bme280MeasurementsReg)})
+	if err != nil {
+		s.logger.Debug("Failed to request temperature")
+	}
+	buffer, err := handle.Read(ctx, 8)
+	if err != nil {
+		return nil, err
+	}
+	if len(buffer) != 8 {
+		return nil, errors.New("i2c read did not get 8 bytes")
+	}
```

Comment:
Again with the design of this sensor. You're sure you want to read then close the handle every-time we call Readings?

Reply:
Multiple devices share the same i2c bus, so we can't just keep the handle open indefinitely.

Reply:
Not what I was saying, you'd open/close in each loop if you poll. But it seems like you don't want to poll it continuously.